core/modules/parser/MySqlListener.h

```diff
@@ -0,0 +1,1679 @@
+// -*- LSST-C++ -*-
+/*
+ * LSST Data Management System
+ * Copyright 2018 LSST Corporation.
+ *
+ * This product includes software developed by the
+ * LSST Project (http://www.lsst.org/).
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+#ifndef LSST_QSERV_PARSER_MYSQLLISTENER_H
+#define LSST_QSERV_PARSER_MYSQLLISTENER_H
+
+#include "parser/MySqlParser.h" // included for contexts. They *could* be forward declared.
+#include "parser/MySqlParserBaseListener.h"
+
+#include <memory>
+#include <stack>
+
+namespace antlr4 {
+    class ParserRuleContext;
+}
+
+namespace lsst {
+namespace qserv {
+namespace query{
+    class SelectStmt;
+}
+namespace parser {
+
+class Adapter;
+class RootAdapter;
+
+class MySqlListener : public MySqlParserBaseListener {
+public:
+    MySqlListener();
+    virtual ~MySqlListener() {}
+
+    // accessors for query types that may have been generated by this listener
+    std::shared_ptr<query::SelectStmt> getSelectStatement() const;
+
+    // This may be raised during listening if there is an error in the enter/exit functions. It may happen
+    // because an unanticipated SQL statement was entered into qserv and we don't yet have the proper
+    // handling for it set up yet.
+    // Any code that calls `listen` on this function should catch this exception, log the `what` message,
+    // and return as though no query was entered.
+    class adapter_order_error : public std::exception {
```

Comment:
I you use `std::runtime_error` as a base then class implementation becomes trivial.